lib/explorer/data_frame.ex

```diff
@@ -104,39 +104,136 @@ defmodule Explorer.DataFrame do
+
+  # The function allows to change the `value` for each pair.
+  defp to_column_pairs(df, pairs, value_fun)
+       when is_column_pairs(pairs) and is_function(value_fun, 1) do
+    existing_columns = names(df)
+
+    pairs
+    |> Enum.map_reduce(nil, fn
+      {column, value}, maybe_map when is_integer(column) ->
+        map = maybe_map || column_index_map(existing_columns)
+
+        existing_column =
+          case Map.fetch(map, column) do
+            {:ok, column} -> column
+            :error -> raise ArgumentError, "column does not exist at index #{column}"
+          end
+
+        {{existing_column, value_fun.(value)}, map}
+
+      {column, value}, maybe_map when is_atom(column) ->
+        column = Atom.to_string(column)
+
+        {{column, value_fun.(value)}, maybe_map}
+
+      {column, value}, maybe_map when is_binary(column) ->
+        {{column, value_fun.(value)}, maybe_map}
+    end)
+    |> then(fn {pairs, _} -> pairs end)
+  end
+
+  defp column_index_map(names),
+    do: for({name, idx} <- Enum.with_index(names), into: %{}, do: {idx, name})
+
+  # Normalize column names without verifying if columns exist.
+  defp to_column_names(names),
+    do: for(name <- Enum.to_list(names), do: to_column_name(name))
+
+  # Normalize column names and raise if column does not exist
+  defp to_existing_columns(df, columns) do
+    existing_columns = names(df)
+
+    columns
+    |> Enum.to_list()
+    |> Enum.map_reduce(nil, fn
+      column, maybe_map when is_integer(column) ->
+        map = maybe_map || column_index_map(existing_columns)
+
+        existing_column =
+          case Map.fetch(map, column) do
+            {:ok, column} -> column
+            :error -> raise ArgumentError, "column does not exist at index #{column}"
+          end
+
+        {existing_column, map}
+
+      column, maybe_map when is_atom(column) ->
+        column = Atom.to_string(column)
+        maybe_raise_column_not_found(existing_columns, column)
+        {column, maybe_map}
+
+      column, maybe_map when is_binary(column) ->
+        maybe_raise_column_not_found(existing_columns, column)
+        {column, maybe_map}
+    end)
+    |> then(fn {columns, _} -> columns end)
+  end
+
   # Access
 
   @behaviour Access
 
   @impl true
-  def fetch(df, columns) when is_list(columns) do
+  def fetch(df, [column | _] = columns) when is_column(column) do
+    columns = to_existing_columns(df, columns)
+
     {:ok, select(df, columns)}
   end
 
-  def fetch(df, columns = %Range{}) do
+  def fetch(df, %Range{} = columns) do
+    columns = to_existing_columns(df, columns)
+
     {:ok, select(df, columns)}
   end
 
-  def fetch(df, column) when is_binary(column) or is_integer(column) do
+  def fetch(df, column) when is_column(column) do
     {:ok, pull(df, column)}
   end
 
   @impl true
-  def pop(df, column) when is_binary(column) or is_integer(column) do
+  def pop(df, column) when is_column(column) do
```

Comment:
We should also support ranges on pop for consistency with fetch. I think we can likely normalize all of these functions.